oauth2/token_hook.go

```diff
@@ -166,12 +196,28 @@ func TokenHook(reg interface {
 			return nil
 		}
 
+		var clientMeta json.RawMessage
+		if hydraClient, ok := requester.GetClient().(*client.Client); ok {
+			clientMeta = json.RawMessage(hydraClient.Metadata)
+		} else {
+			clientMeta = []byte("null")
+		}
+
+		// Always include client authentication parameters, and any sensitive parameters for the requested grant types.
+		sensitiveParameters := append([]string{"client_secret", "client_assertion"}, sensitiveAccessRequestParametersForGrantTypes(requester.GetGrantTypes()...)...)
+
+		payload := maps.Clone(requester.GetRequestForm())
+		// Delete all sensitive parameters from the access request form before sending it to the hook.
+		maps.DeleteFunc(payload, func(s string, _ []string) bool {
+			return slices.Contains(sensitiveParameters, s)
+		})
 		request := Request{
 			ClientID:        requester.GetClient().GetID(),
+			ClientMetadata:  clientMeta,
 			GrantedScopes:   requester.GetGrantedScopes(),
 			GrantedAudience: requester.GetGrantedAudience(),
 			GrantTypes:      requester.GetGrantTypes(),
-			Payload:         requester.Sanitize([]string{"assertion"}).GetRequestForm(),
+			Payload:         payload,
```

Comment:
This is a fundamentally different approach to forwarding request data:

- Before: We restrict the list of parameters to a set of known values that don't leak sensitive data
- Now: We try to guesstimate a list of parameters we better not include but can not guarantee that all requirements are met

I'm really not sure if I agree with this approach. I would rather extend the default allowed parameters to include all known parameters that are OK to forward like `prompt` etc.

The problem with the parameters is that it includes data from the POST payload which is quite likely to contain sensitive things like tokens, passwords, assertions, signatures, ...

Reply:
@aeneasr 

Thanks for having a look, really appreciate your team's work! My goal is to pass any **additional** and non sensitive parameters to the token hook. I was unable to find any guidelines around transmission of additional token endpoint parameters other than that they have to be stored. So I figured this meant they might be fine to utilize in an extensible manner.

For my approach, my thinking was that the set of sensitive parameters for a given token request is actually constrained based on the grant types requested. So its easy to determine this set and filter them out. Is there additional sensitive parameters I didn't include? For what its worth assertion is already specifically leaked to token hooks.

A couple other approaches i considered:
- Always filter all sensitive parameters out but I opted to generate the set based on grant type so there's a clean signal of test failure if a new grant type is added but not supported by this function. 
- Make the list of acceptable additional parameters configurable in the token_hook (with validation to exclude any oauth params).